subscription/subscription.go

```diff
@@ -3,77 +3,91 @@ package subscription
+func NewSubscription(readClient api.Streams_ReadClient, subscriptionId string, eventAppeared chan<- interface {},
+					subscriptionDropped chan<- string) *Subscription {
 	return &Subscription{
 		readClient:          readClient,
 		subscriptionId:      subscriptionId,
 		eventAppeared:       eventAppeared,
-		checkpointReached:   checkpointReached,
 		subscriptionDropped: subscriptionDropped,
+		quit:                make(chan chan error),
 	}
 }
 
 func (subscription *Subscription) Stop() error {
-	subscription.started = false
-	if subscription.subscriptionDropped != nil && !subscription.subscriptionHasBeenDropped {
-		subscription.subscriptionDropped(fmt.Sprintf("User initiated"))
-		subscription.subscriptionHasBeenDropped = true
-	}
-	return subscription.readClient.CloseSend()
+	errc := make(chan error)
+	subscription.quit <- errc
+	return <-errc
 }
 
-func (subscription *Subscription) Start() error {
-	subscription.started = true
-	go func() error {
-		for subscription.started {
-			readResult, err := subscription.readClient.Recv()
-			if err == io.EOF {
-				break
-			}
-			if err != nil {
-				if subscription.subscriptionDropped != nil && !subscription.subscriptionHasBeenDropped {
-					subscription.subscriptionDropped(fmt.Sprintf("Subscription dropped by server: %s", err.Error()))
-					subscription.subscriptionHasBeenDropped = true
+func (subscription *Subscription) Start() {
+	go func() {
+		type recvResult struct {
+			result *api.ReadResp
+			err    error
+		}
+
+		var err error
+		var recvDone chan recvResult
+		var subscriptionHasBeenDropped bool
+		startRecv := time.After(0)
+		for {
+			select {
+			case <-startRecv:
+				recvDone = make(chan recvResult, 1)
+				go func() {
+					result, err := subscription.readClient.Recv()
+					recvDone <- recvResult{result: result, err: err}
+				}()
+			case recvResult := <-recvDone:
+				recvDone = nil
+				result := recvResult.result
+				err = recvResult.err
+				startRecv = time.After(0)
+
+				if err == io.EOF {
+					break
 				}
-				return fmt.Errorf("Failed to perform read. Reason: %v", err)
-			}
-			switch readResult.Content.(type) {
-			case *api.ReadResp_Checkpoint_:
-				{
-					checkpoint := readResult.GetCheckpoint()
-					if subscription.checkpointReached != nil {
-						subscription.checkpointReached(position.Position{
-							Commit:  checkpoint.CommitPosition,
-							Prepare: checkpoint.PreparePosition,
-						})
+				if err != nil {
+					if subscription.subscriptionDropped != nil && !subscriptionHasBeenDropped {
+						subscription.subscriptionDropped <- fmt.Sprintf("Subscription dropped by server: %s", err.Error())
+						subscriptionHasBeenDropped = true
 					}
+					err = fmt.Errorf("Failed to perform read. Reason: %v", err)
 				}
-			case *api.ReadResp_Event:
-				{
-					event := readResult.GetEvent()
-					recordedEvent := event.GetEvent()
-					streamIdentifier := recordedEvent.GetStreamIdentifier()
+				switch result.Content.(type) {
```

Comment:
When the subscription is dropped by the server, I'm getting a null pointer at this line:
```
subscription dropped <-- printed by my client code
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x89af8f]

goroutine 19 [running]:
github.com/EventStore/EventStore-Client-Go/subscription.(*Subscription).Start.func1(0xc000194580)
        /home/shaan/go/src/github.com/EventStore/EventStore-Client-Go/subscription/subscription.go:74 +0x20f
created by github.com/EventStore/EventStore-Client-Go/subscription.(*Subscription).Start
        /home/shaan/go/src/github.com/EventStore/EventStore-Client-Go/subscription/subscription.go:40 +0x3f
```

Adding a `break` statement in the previous `if err != nil {` statement at line 72 fixes it.

Reply:
Added an extra test and fixed the issue.

Reply:
thanks!